### ISSUE TITLE
Cannot use 'uws' transformer with Primus but 'websockets' works fine

### ISSUE THREAD

**Issue by githubaccount624** [uWebSockets/uWebSockets#174]
Hi,

I'm working on a webapp and am currently using Primus with the `'websockets'` transformer. Everything works fine except for server stalling after a couple hundred players.

I'd like to switch to using `'uws'` instead so I can handle more clients but am having trouble getting it working. When I switch the transformer from `'websockets'` to `'uws'` I can still compile the code just fine and there are no errors on the server-side, but I cannot connect from the client anymore.

I checked the developer console and am getting the error:

`'WebSocket connection to 'wss://mywebsite.com/primus?_primuscb=LMt0g9y' failed: Could not decode a text frame as UTF-8.'`

I am using the `'binary'` parser with Primus and thought that maybe `'uws'` didn't support it so I tried changing the parser to `'JSON'` and then got this error on the client:

`WebSocket connection to 'wss://mywebsite.com/primus?_primuscb=LMt1KHS' failed: Received unexpected continuation frame.`

Also, the client tries to keep reconnecting but never succeeds.

Any idea what may be causing this? I would really like to use `'uws'` due to performance reasons but am currently stuck with these errors.

**Comment by alexhultman** [comment]
Do you have any triggering code? When is this happening? What version of Primus / uws? What about non-SSL?

**Comment by alexhultman** [comment]
What version of Primus / uws? What about non-SSL? What do you mean by "stalling"?

**Comment by alexhultman** [comment]
Aha, so the websockets transformer is "stalling" and the uws transformer is completely non-functional at all - you cannot connect one single client?

**Comment by githubaccount624** [comment]
Latest version of primus and uws. Correct, I cannot connect a single client. Didn't test non-ssl since it must be SSL, so if it doesn't work with SSL then I can't use uws.

**Comment by alexhultman** [comment]
So what is "latest" for you?

**Comment by githubaccount624** [comment]
0.7

**Comment by alexhultman** [comment]
Are you running Primus from NPM or from GitHub (master)? The latest Primus release is 3 weeks old so there might very well be incompatibilities. uws itself works with SSL and it passes the Autobahn tests over SSL via Node.js so it would have to be something very strange with something else (like, having an old version of Primus but a new uws version). Do you have any triggering code?

**Comment by alexhultman** [comment]
@lpinca Did you try any SSL with Primus?

**Comment by githubaccount624** [comment]
I'm running Primus from npm. This was happening around a month ago too but I was still in development so figured a newer release by now might have fixed it.

**Comment by githubaccount624** [comment]
SSL works with Primus. I've been using it with `'websockets'` in production for weeks.

**Comment by lpinca** [comment]
@alexhultman no I didn't. primus@5.2.2 is ok, only permessage-deflate support has been added after that.

**Comment by alexhultman** [comment]
I'm running the Autobahn over SSL / Node.js right now (I do this with every release). It works. If you have any simple test Primus code that triggers the failed connection that would be nice. This report is pretty much "SSL is completely broken with Primus / uws" so it should be easy to trigger.

**Comment by alexhultman** [comment]
@lpinca As a side note, I see you missed to pass the 'sec-websocket-extensions' to upgrade. That is needed to negotiate permessage-deflate.

**Comment by lpinca** [comment]
Nice catch.
I wonder if this issue is caused by this [change](https://github.com/primus/primus/commit/667b3577411ae2d77f06e9b431537e33a9c4223e#diff-0ed7397fda91c88e586765cfb21f44d2L54) not being present in primus@5.2.2.

**Comment by alexhultman** [comment]
It could be because if isBuffer evaluates to false it will be read as integer 0 (which is opCode continuation frame like the error). I think "connection error" should be taken with a grain of salt, it could be the first message sent after connection.

**Comment by alexhultman** [comment]
It also makes sense with the UTF-8 because if you have binary data it will evaluate to 1 which is TEXT opCode which will fail since it obviously is not text. I'm pretty sure the issue is a mismatching Primus / uws version. It makes sense.

**Comment by githubaccount624** [comment]
I'm almost 100% sure this is not an SSL issue. I disabled SSL and am still getting this error:

`WebSocket connection to 'ws://mywebsite.com:2585/primus?_primuscb=LMtcOE-' failed: Could not decode a text frame as UTF-8.`

**Comment by alexhultman** [comment]
@githubaccount624 It is a mismatching Primus version.

**Comment by alexhultman** [comment]
If you could clone the master Primus and use it with uws 0.7.0 it should work. I guess there will be a new Primus release sometime soon.

**Comment by githubaccount624** [comment]
Ok, I think I solved it. To anyone reading this who has the same problem, here's what I did to get it working.

1) CD to node_modules
2) git clone https://github.com/primus/primus.git
3) Copy `primus.js` in the node_modules/primus to /dist/primus.js
4) Rename all instances of `primus.use` to `primus.plugin`
5) Rename all instances of `primus.before` to `primus.use`

Recompile. Works for me now.
